README
Mail::MboxParser is a module for working with UNIX-flavoured mailboxes.
This document describes version 0.54.

Its purpose is to _easily_ extract messages from a mailbox with _simple_ 
access to header fields, body and MIME-parts.  It is particularly easy to 
deal with any kind of attached files. See the supplied store_att.pl script 
for an example. 

Prerequisites:
--------------

    + MIME::Tools (probably >= 5.0)
    + MIME::QuotedPrint (don't know which version)
    
    optional:
    + Mail::Mbox::MessageParser (highly recommended but not necessary)
    + URI::Find if you want to use Mail::MboxParser::Mail::Body->extract_urls
    + Encode (standard module for >= perl5.8.0)

Installation:
-------------

Un'tar xfz' the module and 'cd' into the new directory 
Mail-MboxParser-$VERSION.

Then install:
    perl Makefile.PL
    make
    make test
    make install

Win-users having an 'nmake' installed can also do the above, given the
fact they replace 'make' with 'nmake' in the above list.

Contact:
--------
Bugs, feature requests, questions, comments etc. should be directly 
mailed to me.

Tassilo v. Parseval
<tassilo.von.parseval@rwth-aachen.de>



Changelog
v0.54 (Tue Jul 12 09:40:20 CEST 2005)

    + wrong assumptions about quoted-printable encoded
      header-fields were made which lead to attachment 
      filenames often not decoded properly
      (Alper Sari <alper AT sabanciuniv DOT edu>)

v0.53 (Thu Mar 31 09:27:59 CEST 2005)

    + Adds tests for pod-correctness and -coverage

v0.52 (Thu Feb 10 08:21:54 CET 2005)

    + now decodes both base64 and quoted-printable data in the body

v0.51 (Tue Jan  4 17:36:26 CET 2005)

    + Mail::MboxParser::new() didn't check the return value
      of Mail::Mbox::MessageParser's constructor which could lead
      to fatal errors at runtime. When a M::M::MessageParser object
      could not be created, it now silently falls back to the old
      parser.
      (Nigel Horne <njh AT bandsman DOT co DOT uk and
       David Coppit <david AT coppit DOT org>)

v0.50 (Sun Dec 12 09:38:47 CET 2004)

    + At some point in the past this module must have
      stopped working on 5.5.x perls, fixed now

v0.49 (Wed Aug 11 09:43:22 CEST 2004)

    + Dealing with uuencoded MIME parts added + 
      a slight internal clean-up
      (patch by Ed Huott <ed AT supergeek DOT com>)

v0.48 (Wed May 19 07:36:34 CEST 2004)

    + The previous patch was only a partial solution. If an
      attachment wasn't saved, most of the time it meant that
      it wasn't even recognized as such. store_attachment() and
      store_all_attachments() should now find more attachments
      (not more than there are, hopefully;-)
    + Mark Mykkanen <shamu112 AT yahoo DOT com> made me aware that
      the module wouldn't deal correctly with non-seekable filehandles.
      Fixed now
    + Localized *_ properly
    + Replaced 0 with SEEK_SET where it wasn't already used

v0.47 (Mon May  3 08:38:45 CEST 2004)

    + store_attachment/store_all_attachment have new
      parameter 'encoding' to allow filename encoding
      on platforms that require this step (Mac OS X apparently)

v0.46 (Fri Apr  2 09:51:34 CEST 2004)

    + When using the Mail::Mbox::MessageParser,
      the script would die on a file with zero length
      (patch by Lance Cleveland <lance AT proactivewm DOT com>)

v0.45 (Thu Feb  5 10:18:30 CET 2004)

    + Mail::MboxParser::Mail::get_field() wasn't properly
      dealing with multiple occurances of a field when it
      was interleaved with another field
      (patch by Martin Zuziak <zuziak AT fys DOT ku DOT dk>
    + Mail::MboxParser::get_message() and ::next_message() 
      ignored mails with no body

v0.44 (Sat Jan  3 09:40:02 CET 2004)

    + Mail::MboxParser had problems with older versions
      of FileHandle::Unget. Makefile.PL now checks
      that a useable version is installed

v0.43 (Sat Oct 18 16:23:29 CEST 2003)

    + Return values of File::Temp::tempfile() were 
      mistakenly swapped 
      (patch by Lyon Lemmens <lyon DOT lemmens AT oracle DOT com>)

v0.42 (Tue Oct 14 10:46:11 CEST 2003)

    + fixed "Call SETUP_CACHE() before calling new()..." warning
      injected by Mail::Mbox::MessageParser
    + hopefully removed function-redefinition warnings

v0.41 (Thu Sep  4 09:52:04 CEST 2003)

    + forgot to import tempfile() from File::Temp:
      this resulted in an
      "Undefined subroutine &Mail::MboxParser::tempfile
      called at..."
    + wrong SYNOPSIS: store_all_attachments() receives key/value 
      pairs as arguments

v0.40 (Tue Aug 19 09:14:08 CEST 2003)

    + next_message() and get_message() dropped last
      line of body if it did not end with a newline
      (spotted by <marco AT aktex DOT net>)

v0.39 (Tue Aug  5 08:06:50 CEST 2003)

    + now uses David Coppit's new Mail::Mbox::MessageParser module 
      which provides much faster and more robust parsing 
      (the old parser can still be used (see docs))
      !! therefore 0.39 should probably be marked as beta !!
    + removed the silly benchmark results from the PODs
      (they are no longer up-to-date)

v0.38 (Wed Jan 15 14:34:49 CET 2003)

    + fixed wrong splittal: in to(), cc(), from()
      '"name, first" <email@host.com>' would be split even though
      the comma is within double quotes
      (patched by Robert McArthur <mcarthur AT dstc DOT edu DOT au>)
    + in same methods:
      'name' is set to "First Name" if it was previously empty and
      the email address looks like "first.name@email.com"
      (suggested and patched by Robert McArthur)
    + fixed a mysterious "Uninitialized value in list assignment" 
      warning in Body.pm
    + localized a filehandle inside Mail.pm
    + fixed typo in Makefile.PL

v0.37 (Tue Nov 26 10:10:18 CET 2002)

    + fixed some inconsistencies when constructing a 
      Mail::MboxParser::Mail object manually
      (thanks to Premjit Nair <premjit AT dpa DOT co DOT ae> for tracking 
       that down)
    + some of the PODs in Mail::MboxParser::Mail were really horrible,
      a lot of minor errors along with bad formatting fixed

v0.36 (Sun Nov 24 13:24:59 CET 2002)

    + Mail::MboxParser::Mail::Body did not overload stringification
      (patched by Steven W McDougall <swmcd AT TheWorld DOT com>)

v0.35 (Thu Aug 22 08:54:01 CEST 2002)

    + 0.34 forgot to bump the version number

v0.34 (Sun Aug 11 16:25:53 CEST 2002)

    + new option to strip signature on output
      (implicitely suggested by David Bryson)
    + on unfolding of continuation header-lines,
      whitespaces disappeared
      (fix by Nathan Uno (<Nathan AT unos DOT net>)
    + added MIME::QuotedPrint to prerequisites
    
v0.33 (Thu Jun 20 07:26:25 CEST 2002)

    This is a bug-fix only release:
    + multiple occurances of a header-field were
      incorrectly handled: only the last occurance
      was returned, now you get an array_ref
    + localizing of file-handles
    + fixed $msg->as_string (broken in 0.32)
    + store_attachment() and store_all_attachments()
      fixed to work without the 'store_only' argument
    + list/scalar context issue fixed in get_entities(),

MboxParser/Base.pm
# Mail::MboxParser - object-oriented access to UNIX-mailboxes
# base-class for all other classes in Mail::MboxParser
#
# Copyright (C) 2001  Tassilo v. Parseval
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.

# Version: $Id: Base.pm,v 1.6 2002/02/21 09:06:14 parkerpine Exp $

package Mail::MboxParser::Base;

require 5.004;

use strict;
use vars qw($VERSION);
$VERSION = "0.07";

sub new(@) {
    my ($class, @args) = @_;

    if ($class eq __PACKAGE__) {
	use Carp;
	my $package = __PACKAGE__;
	croak <<USAGE;
$package should not really be instantiated directly. 
Instead, create one of its derived subclasses such as Mail::MboxParser.
USAGE
    }
    my $self = bless {}, $class;
    $self->init(@args);
}

sub error() { shift->{LAST_ERR} }

sub log() { shift->{LAST_LOG} }

sub reset_last { 
    my $self = shift;
    ($self->{LAST_ERR}, $self->{LAST_LOG}) = (undef, undef);
}

1;

__END__

=head1 NAME

Mail::MboxParser::Base - base clase for all other classes

=head1 DESCRIPTION

Nothing to describe nor to document here. Read L<Mail::MboxParser> on how to use the module.

=head1 VERSION

This is version 0.54.

=head1 AUTHOR AND COPYRIGHT

Tassilo von Parseval <tassilo.von.parseval@rwth-aachen.de>

Copyright (c)  2001-2005 Tassilo von Parseval.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


MboxParser/Mail/Convertable.pm
# Mail::MboxParser - object-oriented access to UNIX-mailboxes
# Convertable.pm   - allow altering of mail for multiple purposes
#
# Copyright (C) 2001  Tassilo v. Parseval
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.

# Version: $Id: Convertable.pm,v 1.6 2002/02/21 09:06:15 parkerpine Exp $

package Mail::MboxParser::Mail::Convertable;

require 5.004;

use Carp;

use strict;
use base qw(Exporter);
use vars qw($VERSION @EXPORT @ISA $AUTOLOAD); 
$VERSION 	= "0.06";
@EXPORT  	= qw();
@ISA	 	= qw(Mail::MboxParser::Base Mail::MboxParser::Mail);

sub init(@) {	
	my ($self, $ent, @args) = @_;
	$self->{TOP_ENTITY}	    = $ent;
	$self;
}

sub delete_from_header(@) { 
    my $self = shift;
    $self->{TOP_ENTITY}->head->delete($_) for @_; 
}

sub add_to_header(@) { 	
	my ($self, $what) = (shift, shift);

	if (not ref $what) {
		croak <<EOC;
Error: First argument to add_to_header must be a reference to a list with 
two elements.
EOC
	}
	my %args = @_;
	my $index;
	$args{where} = 'BEHIND' if not exists $args{where};
	if ($args{where} eq 'BEFORE') { $index = 0 }
	if ($args{where} eq 'BEHIND') { $index = -1 }
	
	$self->{TOP_ENTITY}->head->add(@{$what});
}

sub replace_in_header($$) {
	if (@_ != 3) {
		croak <<EOC;
Error: replace_in_headers() needs two arguments.
EOC
	}
	shift->{TOP_ENTITY}->head->replace(shift, shift);
}



1;

__END__

=head1 NAME

Mail::MboxParser::Mail::Convertable - convert mail for sending etc.

=head1 SYNOPSIS

    use Mail::MboxParser;

    [...]
    
    # $msg is a Mail::MboxParser::Mail-object
    my $mail = $msg->make_convertable; 

    $mail->delete_from_header('date', 'message-id');
    $mail->replace_in_header('to', 'john.doe@foobar.com');
    $mail->add_to_header( ['cc', 'john.does.brother@foobar.com'],
                          where => 'BEHIND' );
    $mail->send('sendmail');
    

=head1 DESCRIPTION

This class adds means to convert an email object into something that could be send via SMTP, NNTP or dumped to a file or filehandle. Therefore, methods are provided that change the structure of an email which includes adding and removing of header-fields, MIME-parts etc and transforming them into objects of related modules.

Currently, only basic manipulation of the header and sending using Mail::Mailer is provided. More is to come soon.

This class works non-destructive. You first create a Convertable-object and do any modifications on this while the Mail-object from which it was derived will not be touched.

=head1 METHODS

=over 4

=item delete_from_header(header-fields)

Given a list of header-field names, these fields will be removed from the header. If you want to re-send a message, you could for instance remove the cc-field cause otherwise the message would be carbon-copied to the addresses listed in the cc-field.

=item add_to_header(array-ref)

=item add_to_header(array-ref, where => 'BEFORE' | 'BEHIND')

add_to_header() takes a reference to a two-element list whose first element specifies the header-field to add or to add to while the second elements specifies the data that should be added. 'where' specifies whether to add at the beginning or at the end of the header. Defaults to 'BEHIND' if not given.

=item replace_in_header(header-field, new_data)

First element must be the header-field to be replaced while the second argument must be a string indicating what will be the new content of the header-field.

=item send(command, args)

Literally inherited from Mail::Internet. Commands can be "mail" (using the UNIX-mail program), "sendmail" (using a configured sendmail or compatible MTA like exim), "smtp" (for using Net::SMTP) and "test" which will only display what would be sent using /bin/echo. Additional arguments will be passed on to Mail::Mailer->new() which is in fact what Mail::Internet->send() uses.

For more details, see L<Mail::Mailer>

=back 

=head1 VERSION

This is version 0.54.

=head1 AUTHOR AND COPYRIGHT

Tassilo von Parseval <tassilo.von.parseval@rwth-aachen.de>

Copyright (c)  2001-2005 Tassilo von Parseval.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<Mail::Internet>, L<Mail::Mailer>



MboxParser/Mail/Body.pm
# Mail::MboxParser - object-oriented access to UNIX-mailboxes
# Body.pm		   - the (textual) body of an email
#
# Copyright (C) 2001  Tassilo v. Parseval
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.

# Version: $Id: Body.pm,v 1.14 2002/02/21 09:06:14 parkerpine Exp $

package Mail::MboxParser::Mail::Body;

require 5.004;

use Carp;

use strict;
use base qw(Exporter);
use vars qw($VERSION @EXPORT @ISA $AUTOLOAD $_HAVE_NOT_URI_FIND);
$VERSION 	= "0.15";
@EXPORT  	= qw();
@ISA	 	= qw(Mail::MboxParser::Base Mail::MboxParser::Mail);

use overload '""' => sub { shift->as_string }, fallback => 1;

BEGIN { 
    eval { require URI::Find; };
    if ($@) { 
	$_HAVE_NOT_URI_FIND = 1;
    }
}

sub init(@) {
    my ($self, $ent, $bound, $conf) = @_;
    $self->{CONTENT}	= $ent->body; 
    $self->{BOUNDARY}	= $bound;	     # the one in Content-type
    $self->{ARGS}	= $conf;

    $self->{ARGS}->{decode} ||= 'NEVER';

    $self->_make_decoder($ent->head->mime_encoding) 
	if $self->{ARGS}->{decode} =~ /BODY|ALL/;;
    $self;
}

sub _make_decoder {
    my ($self, $enc) = @_;
    if ($enc eq 'base64') {
	require MIME::Base64;
	return $self->{DECODER} = sub { MIME::Base64::decode_base64(shift) };
    } 
    if ($enc eq 'quoted-printable') {
	require MIME::QuotedPrint;
	return $self->{DECODER} = sub { MIME::QuotedPrint::decode_qp(shift) };
    }
    $self->{DECODER} = sub { $_[0] };
} 

sub as_string {	
    my ($self, %args) = @_;
    $self->reset_last;
    return join "", $self->as_lines(strip_sig => 1) if $args{strip_sig};
    my $decode = $self->{ARGS}->{decode};
    if ($decode eq 'BODY' || $decode eq 'ALL') {
	return join "", map { $self->{DECODER}->($_) } @{$self->{CONTENT}};
    }
    return join "", @{$self->{CONTENT}};
}
	

sub as_lines() { 
    my ($self, %args) = @_;
    $self->reset_last;
    my $decode = $self->{ARGS}->{decode};
    if ($decode eq 'BODY' || $decode eq 'ALL') {
	return map { $self->{DECODER}->($_) } @{$self->{CONTENT}};
    }

    return @{$self->{CONTENT}} if ! $args{strip_sig};

    my @lines;
    for (@{ $self->{CONTENT} }) {
	last if /^--\040?[\r\n]?$/;
	push @lines, $_;
    }
    return @lines;
}
					   
	
sub signature() {
    my $self = shift;
    $self->reset_last;
    my $decode = $self->{ARGS}->{decode};
    my $bound = $self->{BOUNDARY};

    my @signature;
    my $seperator = 0;
    for (@{$self->{CONTENT}}) {

	# we are still outside the signature
	if (! /^--\040?[\r\n]?$/ && not $seperator) {
	    next;
	}

	# we hit the signature delimiter (--)
	elsif (not $seperator) { $seperator = 1; next }

	chomp;	

	# we are inside signature: is line perhaps MIME-boundary?
	last if $bound && /^--\Q$bound\E/ && $seperator;

	# none of the above => signature line
	push @signature, $_; 
    }
	
    $self->{LAST_ERR} = "No signature found" if !@signature;
    if ($decode eq 'BODY' || $decode eq 'ALL') {
	$_ = $self->{DECODER}->($_) for @signature;
    }
    return @signature if $seperator;
    return ();
}

sub extract_urls(@) {
    my ($self, %args) = @_;
    $self->reset_last;

    $args{unique} = 0 if not exists $args{unique};

    if ($_HAVE_NOT_URI_FIND) {
	carp <<EOW;
You need the URI::Find module in order to use extract_urls.
EOW
	return;
    }
    else { 
	my @uris; my %seen;

	for my $line (@{$self->{CONTENT}}) {
	    chomp $line;
	    URI::Find::find_uris($line, sub {
		    my (undef, $url) = @_;
		    $line =~ s/^\s+|\s+$//;
		    if (not $seen{$url}) {
			push @uris, { url => $url, context => $line };
		    }
		    $seen{$url}++ if $args{unique};
		}
	    );
	}
	$self->{LAST_ERR} = "No URLs found" if @uris == 0;

	return @uris;
    }
}

sub quotes() {
    my $self = shift;
    my $decode = $self->{ARGS}->{decode};
    $self->reset_last;

    my %ret;
    my $q 		= 0; # num of '>'
    my $in 		= 0; # being inside a quote
    my $last 	= 0; # num of quotes in last line

    for (@{$self->{CONTENT}}) {

	if ($decode eq 'ALL' || $decode eq 'BODY') {
	    $_ = $self->{DECODER}->($_);
	}
        
	# count quotation signs
	$q = 0;
	my $t = "a" x length;
	for my $c (unpack $t, $_) {
	    if ($c eq '>') 		{ $q++ }
	    if ($c ne '>' && $c ne ' ') { last }
	}

	# first: create a hash-element for level $q
	if (! exists $ret{$q}) {
	    $ret{$q} = [];
	}

	# if last line had the same level as current one:
	# attach the line to the last one
	if ($last == $q) {
	    if (@{$ret{$q}} == 0)   { $ret{$q}->[$q] .= $_ }
	    else		    { $ret{$q}->[-1] .= $_ }
	}
	# if not:
	# create a new array-element in the appropriate hash-element
	else { 
	    push @{$ret{$q}}, $_;
	}
	$last = $q;
    }
    return \%ret;
}


1;

__END__

=head1 NAME

Mail::MboxParser::Mail::Body - rudimentary mail-body object

=head1 SYNOPSIS

    use Mail::MboxParser;

    [...]

    # $msg is a Mail::MboxParser::Mail
    my $body = $msg->body(0);

    # or preferably

    my $body = $msg->body($msg->find_body);

    for my $line ($body->signature) { print $line, "\n" }
    for my $url ($body->extract_urls(unique => 1)) {
        print $url->{url}, "\n";
        print $url->{context}, "\n";
    }
        
=head1 DESCRIPTION

This class represents the body of an email-message.  Since emails can have
multiple MIME-parts and each of these parts has a body it is not always easy to
say which part actually holds the text of the message (if there is any at all).
Mail::MboxParser::Mail::find_body will help and suggest a part.

=head1 METHODS

=over 4

=item B<as_string ([strip_sig =E<gt> 1])>

Returns the textual representation of the body as one string. Decoding takes
place when the mailbox has been opened using the decode => 'BODY' | 'ALL'
option.

If 'strip_sig' is set to a true value, the signature is stripped from the
string.

=item B<as_lines ([strip_sig =E<gt> 1])>

Sames as as_string() just that you get an array of lines with newlines attached
to each line.

B<NOTE:> When the body is actually some encoded binary data (most commonly such
a body is base64-encoded), you can still use this method. Then you wont really
get proper lines. Instead you get chunks of binary data that you should
concatenate as in

    my $binary = join "", $body->as_lines;

If 'strip_sig' is set to a true value, the signature is stripped from the
string.

=item B<signature>

Returns the signature of a message as an array of lines. Trailing newlines are
already removed.

$body->error returns a string if no signature has been found.

=item B<extract_urls>

=item B<extract_urls (unique =E<gt> 1)>

Returns an array of hash-refs. Each hash-ref has two fields: 'url' and
'context' where context is the line in which the 'url' appeared.

When calling it like $mail->extract_urls(unique => 1), duplicate URLs will be
filtered out regardless of the 'context'. That's useful if you just want a list
of all URLs that can be found in your mails.

$body->error() will return a string if no URLs could be found within the body.

=item B<quotes>

Returns a hash-ref of array-refs where the hash-keys are the several levels of
quotation. Each array-element contains the paragraphs of this quotation-level
as one string. Example:

	my $quotes = $msg->body($msg->find_body)->quotes;
	print $quotes->{1}->[0], "\n";
	print $quotes->{0}->[0], "\n";

This should print the first paragraph of the mail-body that has been quoted
once and below that the paragraph that supposedly is the reply to this
paragraph. Perhaps thus:

	> I had been trying to work with the CGI module 
	> but I didn't yet fully understand it.

	Ah, it is tricky. Have you read the CGI-FAQ that 
	comes with the module?

Mark that empty lines will not be ignored and are part of the lines contained
in the array of $quotes->{0}.

So below is a little code-snippet that should, in most cases, restore the first
5 paragraphs (containing quote-level 0 and 1) of an email:

	for (0 .. 4) {
		print $quotes->{0}->[$_];
		print $quotes->{1}->[$_];
	}

Since quotes() considers an empty line between two quotes paragraphs as a
paragraph in $quotes->{0}, the paragraphs with one quote and those with zero
are balanced. That means: 

scalar @{$quotes->{0}} - DIFF == scalar @{$quotes->{1}} where DIFF is element
of {-1, 0, 1}.

Unfortunately, quotes() can up to now only deal with '>' as quotation-marks.

=back

=head1 VERSION

This is version 0.54.

=head1 AUTHOR AND COPYRIGHT

Tassilo von Parseval <tassilo.von.parseval@rwth-aachen.de>

Copyright (c)  2001-2005 Tassilo von Parseval.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

=cut
